### PR TITLE
Close resources open for subprocess.Popen and open

### PR DESCRIPTION
**Pull request by kinow**
An example PR for #3264. But having changes in the scheduler, the pool of processes, and other areas that are still a bit dark to me, I think this will be more a reference for @hjoliver, @matthewrmshin or anybody with more experience with Cylc internals and that may be able to work on that issue :+1: 

**Requirements check-list**
- [X] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [X] Contains logically grouped changes (else tidy your branch by rebase).
- [X] Does not contain off-topic changes (use other PRs for other changes).

- [ ] Appropriate tests are included (unit and/or functional).
- [ ] These changes are already covered by existing tests.

- [ ] Includes an appropriate entry in the release change log `CHANGES.md`.
- [x] No change log entry required (e.g. change is small or internal only).

- [X] No documentation update required for this change.

### CODE REVIEW

**Comment by kinow** [comment]
Didn't profile Cylc, but in theory closing the file descriptors sooner instead of leaving them for Python to take care alter could improve memory I guess (and avoid running out of file descriptors, though it's very unlikely I think).

File: cylc/flow/batch_sys_handlers/background.py [hunk 1]
```diff
@@ -68,7 +69,7 @@ def submit(cls, job_file_path, submit_opts):
                     job_file_path,
                 ],
                 preexec_fn=os.setpgrp,
-                stdin=open(os.devnull),
+                stdin=DEVNULL,
```
Comment (matthewrmshin) [inline cylc/flow/batch_sys_handlers/background.py]: Very nice! (New in Python 3.3.)

File: cylc/flow/batch_sys_manager.py [hunk 2]
```diff
@@ -634,6 +634,7 @@ def _job_submit_impl(
         # Submit job
         batch_sys = self._get_sys(batch_sys_name)
         proc_stdin_arg = None
+        # TODO: close stdin
```
Comment (matthewrmshin) [inline cylc/flow/batch_sys_manager.py]: Can do this after line 678?

File: cylc/flow/cylc_subproc.py [hunk 3]
```diff
@@ -33,7 +33,8 @@ def procopen(cmd, bufsize=0, executable=None, stdin=None, stdout=None,
              stderr=None, preexec_fn=None, close_fds=False, usesh=False,
              cwd=None, env=None, universal_newlines=False, startupinfo=None,
              creationflags=0, splitcmd=False, stdoutpipe=False,
-             stdoutout=False, stderrpipe=False, stderrout=False):
+             stdoutout=False, stderrpipe=False, stderrout=False,
+             stdindevnull=False):
```
Comment (matthewrmshin) [inline cylc/flow/cylc_subproc.py]: In our usages, STDIN should probably default to `/dev/null` for safety. There are only very small number of cases where we actually want to pipe the *current* STDIN to the child command.

Comment (kinow) [reply cylc/flow/cylc_subproc.py]: Got it, changed :+1:

File: cylc/flow/batch_sys_manager.py (hunk 2 above)
Comment (kinow) [reply cylc/flow/batch_sys_manager.py]: Actually I had done it. I think I managed to eliminate the TODOs, except for one... then I reverted for not being brave enough to change, and not knowing how exactly to test it and make sure it didn't break anything.

Not entirely sure what code in `batch_sys_manager.py` is tested in Travis, and what part needs to invoke something like PBS/etc? But might be worth and check coverage in Codecov. Changed! :+1:

Comment (kinow) [reply cylc/flow/batch_sys_manager.py]: ~Changed~, scratch that. Still not confident in changing this :confused: 

If you look at the `if` statement right after this `TODO`, if I am reading it right, if the `batch_sys` has the attribute `get_submit_stdin`, then it looks like the value of `proc_stdin_value` is changed... so we may not have a resource in 678? And may instead have an `str` I guess?

Comment (matthewrmshin) [reply cylc/flow/batch_sys_manager.py]: Sorry, looks like `.get_submit_stdin` is not documented properly. A quick `git grep` suggests that it is used by the `at` and `lsf` handlers.

Looking at the returned pair `proc_stdin_arg, proc_stdin_value`. The  `proc_stdin_arg` should be an opened file handle, `subprocess.PIPE` or `None`. (Not sure why it is tested for `str`.) The `proc_stdin_value` is either `None` or a `str` containing the content to pipe to STDIN.

Comment (kinow) [reply cylc/flow/batch_sys_manager.py]: Huh, confusing. I think I will probably leave the TODO for now (if you would like to take over this PR and add/edit commits, feel free to push any changes here or start a new one)

Comment (matthewrmshin) [reply cylc/flow/batch_sys_manager.py]: OK. I'll raise a PR against your branch.

Comment (matthewrmshin) [reply cylc/flow/batch_sys_manager.py]: kinow/cylc-flow#4

**Comment by matthewrmshin** [comment]
Hi @kinow I guess you are still keen to write some new tests for this change? (Note: all of these are very well covered by the original test battery.)

**Comment by kinow** [comment]
>Hi @kinow I guess you are still keen to write some new tests for this change? (Note: all of these are very well covered by the original test battery.)

I also thought the test battery would cover this. Except for batch sys handler, which I am not sure how to test (don't know that part of the code well enough).

I'm fine if you have a look and can confirm the change doesn't seem to have anything strange, and merge as-is.

**Comment by matthewrmshin** [comment]
The changes under `cylc.flow.batch_sys_manager` and `cylc.flow.batch_sys_handlers.background` are covered by well covered by the test battery. The changes in the former are covered by usage of `at` for job submission. The changes in the latter are covered by any background job submission (which is the default for most suite tasks in the test battery).

**Comment by kinow** [comment]
Got it @matthewrmshin , moved from draft to ready for review (as I was not sure others had access for doing this? but I think we agree on this current version)

**Comment by hjoliver** [comment]
So we've got two approvals, and I think the consensus is there's no need for additional tests.  Is that right @kinow?  If so, we can merge this.

**Comment by kinow** [comment]
>So we've got two approvals, and I think the consensus is there's no need for additional tests. Is that right @kinow? If so, we can merge this.

Yup! :rocket:

**Pull request merged**